msrestazure/azure_local_creds_prober.py

```diff
@@ -0,0 +1,142 @@
+        self.creds = creds
+        if not subscription_id:
+            try:
+                from azure.mgmt.resource.subscriptions import SubscriptionClient
+                subscriptions = list(SubscriptionClient(creds).subscriptions.list())
+                if subscriptions:
+                    subscription_ids = [s.id.split('/')[-1] for s in subscriptions]
+                    subscription_id = subscription_ids[0]
+                    _LOGGER.warning('Found subscription "%s" to use', subscription_ids[0])
+                    if len(subscription_ids) > 1:
+                        _LOGGER.warning('You also have accesses to a few other subscriptions "%S".'
+                                        ' You can supply subscription_id on creating the probe object')
+            except ImportError:  # should be rare
+                _LOGGER.warning('Failed to load azure.mgmt.resource.subscriptions to find the default subscription.'
+                                ' If this is expected, supply subscription_id on creating the probe object')
+        self.subscription_id = subscription_id
+
+
+class CLICredentials(BasicTokenAuthentication):
+
+    def __init__(self, resource=None, subscription_id=None):   # allow subscriptions
+        super(CLICredentials, self).__init__(None)
+        uname = platform.uname()
+        # python 2, `platform.uname()` returns: tuple(system, node, release, version, machine, processor)
+        platform_name = getattr(uname, 'system', None) or uname[0]
+        platform_name = platform_name.lower()
+        if platform_name == 'windows':
+            program_files_folder = os.environ.get('ProgramFiles(x86)') or os.environ.get('ProgramFiles')
+            probing_paths = [os.path.join(program_files_folder, 'Microsoft SDKs', 'Azure', 'CLI2', 'wbin', 'az.cmd')]
+        else:
+            probing_paths = ['/usr/bin/az', '/usr/local/bin/az']
+        
+        cli_path = next((p for p in probing_paths if os.path.isfile(p)), None)
+        if cli_path is None:
+            raise NotImplementedError('Azure CLI is not installed')
+        self.cli_path = cli_path
+        self.resource = resource
+        self.subscription_id = subscription_id
+        self.token = None
+
+    def set_token(self):
+        args = [self.cli_path, 'account', 'get-access-token']
+        if self.subscription_id:
+            args.extend(['--subscription', self.subscription_id])
+        p = Popen(args, stdout=PIPE, stderr=PIPE)
+        stdout, stderr = p.communicate()
+        p.wait()
+        if stderr:
+            raise ValueError('Retrieving acccess token failed: ' + stderr)
+        info = json.loads(stdout)
+        self.scheme, self.token = info['tokenType'], {'access_token': info['accessToken']}
+
+    def signed_session(self, session=None):
+        # Token cache is handled by the VM extension, call each time to avoid expiration
+        self.set_token()
```

Comment:
Should, the CLI be called each request to refresh the token?

Reply:
I assume you meant sub-shelling is kind of slow? Yes, I can optimize by leveraging the `expiresOn` information returned from the `get-access-token` command